README.md

# podoff

[![Build Status](https://secure.travis-ci.org/jmettraux/podoff.png)](http://travis-ci.org/jmettraux/podoff)
[![Gem Version](https://badge.fury.io/rb/podoff.png)](http://badge.fury.io/rb/podoff)

A Ruby tool to deface PDF documents.

Uses "incremental updates" to do so.

Podoff is used to write over PDF documents. Those documents should first be uncompressed (and recompressed) (how? see [below](#preparing-documents-for-use-with-podoff))

```ruby
require 'podoff'

d = Podoff.load('d2.pdf')
  # load my d2.pdf

fo = d.add_base_font('Helvetica')
  # make sure the document knows about "Helvetica"
  # (one of the base 13 or 14 fonts PDF readers know about)


pa = d.page(1)
  # grab first page of the document

pa.insert_font('/MyHelvetica', fo)
  # link "MyHelvetica" to the base font above for this page

st =
  d.add_stream {
    tf '/MyHelvetica', 12 # Helvetica size 12
    bt 100, 100, "#{Time.now} stamped via podoff" # text at bottom left
  }

pa.insert_content(st)
  # add content to page

d.write('d3.pdf')
  # write stamped document to d3.pdf
```

For more about the podoff "api", read ["how I use podoff"](#how-i-use-podoff).

If you're looking for serious libraries, look at:

* https://github.com/payrollhero/pdf_tempura
* https://github.com/prawnpdf/prawn-templates

And more recently:

* https://github.com/gettalong/hexapdf


## preparing documents for use with podoff

Podoff is naive and can't read xref tables in object streams. You have to work against PDF documents that have vanilla xref tables. [Qpdf](http://qpdf.sourceforge.net/) to the rescue.

Given a doc0.pdf you can produce such a document by doing:
```
qpdf --object-streams=disable doc0.pdf doc1.pdf
```
doc1.pdf is now ready for overwriting with podoff.

qpdf has rewritten the PDF, extracting the xref table but keeping the streams compressed.


## bin/podoff

`bin/podoff` is a command-line tool for to preparing/check PDFs before use.

```
$ ./bin/podoff -h

Usage: ./bin/podoff [option] {fname}

    -o, --objs                       List objs
    -w, --rewrite                    Rewrite
    -s, --stamp                      Apply time stamp at bottom of each page
    -r, --recompress                 Recompress
    --version                        Show version
    -h, --help                       Show this message
```

`--recompress` is mostly an alias for `qpdf --object-streams=disable in.pdf out.pf`

`--stamp` is used to check whether podoff can add a time stamp on each page of an input PDF.


## how I use podoff

In the application which necessitated the creation of podoff, there are two PDF to generate from time to time.

I keep those two PDFs in memory.

```ruby
# lib/myapp/pdf.rb

require 'podoff'

module MyApp::Pdf

  DOC0 = Podoff.load('pdf_templates/d0.pdf')
  DOC1 = Podoff.load('pdf_templates/d1.pdf')

  def generate_doc0(data, path)

    d = DOC0.dup # shallow copy of the document
    d.add_fonts

    pa2 = d.page(2)
    st = d.add_stream # open stream...

    st.font 'MyHelv', 12 # font is an alias to tf
    st.color 0, 0, 0 # rgb black
    st.text 100, 100, data['customer_name']
    st.text 100, 80, data['customer_phone']
    st.text 100, 60, data['date'] if data['date']
      # fill in customer info on page 2

    pa2.insert_content(st) ... close stream (yes, you can use a block too)

    pa3 = d.page(3)
    pa3.insert_content(d.add_stream { check 52, 100 }) if data['discount']
      # a single check on page 3 if the customer gets a discount

    d.write(path)
  end

  # ...
end

module Podoff # adding a few helper methods to the podoff classes

  class Document

    # Makes sure Helvetica and ZapfDingbats are available
    # on each page of the document
    #
    def add_fonts

      fo0 = add_base_font('/Helvetica')
      fo1 = add_base_font('/ZapfDingbats')
      fo2 = add_base_font('/Courier')

      pages.each { |pa|
        pa = re_add(pa)
        pa.insert_font('/MyHelv', fo0)
        pa.insert_font('/MyZapf', fo1)
        pa.insert_font('/MyMono', fo2)
      }
    end
  end

  class Stream

    # Places a check mark ✓ at x, y
    #
    def check(x, y)

      font = @font            # save current font
      self.tf '/MyZapf', 12   # switch to ZapfDingbats size 12
      self.bt x, y, '3'       # check mark
      @font = font            # get back to saved font
    end
  end
end
```

The documents are kept in memory, as generation request comes, the get duplicated, incrementally updated and the filled documents are written to disk. The duplication doesn't copy the whole document file, only the references to the "obj" in the document get copied.

### Podoff::Document

```ruby
class Podoff::Document

  def self.load(path, encoding='iso-8859-1')
    # Podoff.load(path, encoding) is a shortcut to this method

  def dup
    # Makes a shallow copy of the document

  def add_base_font(name)
    # Given a name in the base 13/14 fonts readers are supposed to know,
    # ensures the document has access to the font.
    # Usually "Helvetica" or "ZapfDingbats".

  def pages
    # Returns an array of all the objs that are pages

  def page(index)
    # Starts at 1, returns a page obj. Understands negative indexes, like
    # -1 for the last page.

  def add_stream(src=nil, &block)
    # Prepares a new obj with a stream
    # If src is given places the src string in the stream.
    # If a block is given executes the block in the context of the
    # Podoff::Stream instance.
    # If no src and no block, simply returns the Podoff::Stream wrapped inside
    # of the new obj (see example code above)

  def re_add(obj_or_ref)
    # Given an obj or a ref (like "1234 0") to an obj, copies that obj
    # and re-adds it to the document.
    # This is necessary for the incremental updates podoff uses, if you add
    # an obj to the Contents list of a page, you have to add it to the
    # re-added page, not directly to the original page.

  def write(path=:string)
    # Writes the document, with incremental updates to a file given by its path.
    # If the path is :string, will simply return the string containing the
    # whole document

  def rewrite(path=:string)
    # Like #write, but squashes the incremental updates in the document.
    # Takes more time and memory and might fail (remember, podoff is very
    # naive (as his author is)). Test with care...

  #
  # a bit lower-level...

  def objs
    # returns the hash { String/obj_ref => Podoff::Obj/obj_instance }
```

### Podoff::Obj

A PDF document is mostly a hierarchy of `obj` elements. `Podoff::Obj` points to such elements (see `Podoff::Document#objs`).

```ruby
class Podoff::Obj

  def insert_font(font_nick, font_obj_or_ref)
  def insert_contents(obj_or_ref)
```

### Podoff::Stream

Here are the methods used to write to streams.

```ruby
class Podoff::Stream

  # set the current font and font size for the stream
  #
  def tf(font_name, font_size)
  alias font tf

  # set the current color for the stream
  #
  def rg(red, green, blue)
  def rg(string) # like '#fffff' or 'white'
  alias rgb rg
  alias color rg

  # write a piece of text at a given position
  #
  def bt(x, y, text)
  def bt(x, y, text, rgb: 'blue')
  alias text bt

  # draw a rectangle
  #
  def re(x, y, *a)
  alias rect re
  alias rectangle re
    #
  st.re(10, 20, 30, 40, rgb: [ 0.0, 0.0, 0.0 ])
  st.rect(11, 21, w: 31, h: 41, rgb: [ 0.1, 0.1, 0.1 ])
  st.rectangle(12, 22, 32, 42, rgb: [ 0.2, 0.2, 0.2 ])
  # ...

  # draw a line
  #
  def line(x0, y0, x1, y1, *a)
    #
  st.line(1, 1, 2, 2)
  st.line(1, 1, 2, 2, 3, 3)
  st.line([ 1, 1 ], [ 2, 2 ], [ 3, 3 ])
  st.line([ 1, 1 ], [ 2, 2 ], [ 3, 3 ], rgb: [ 0.5, 0.5, 0.5 ])
  st.line(1, 1, 2, 2, rgb: [ 0.7, 0.7, 0.7 ])
  # ...
  st.line(1, 1, 2, 2, rgb: [ 0.7, 0.7, 0.7 ], width: 0.3)
  # ...
```


## disclaimer

The author of this tool/library have no link whatsoever with the authors of the sample PDF documents found under `pdfs/`. Those documents have been selected because they are representative of the PDF forms podoff is meant to ~~deface~~fill.


## known bugs

* podoff parsing is naive, documents that contain uncompressed streams with "endobj", "startxref", "/Root" will disorient podoff
* completely candid about encoding (only used it for British English documents so far)


## links

* http://qpdf.sourceforge.net/ source: https://github.com/qpdf/qpdf

* http://www.slideshare.net/ange4771/advanced-pdf-tricks


## LICENSE

MIT, see [LICENSE.txt](LICENSE.txt)



CHANGELOG.md

# CHANGELOG.md


## podoff 1.4.0  released 2020-04-06

* Introduce Stream#deflate=bool


## podoff 1.3.0  released 2020-03-30

* Introduce Podoff::Stream#line
* Introduce Podoff::Stream#rectangle


## podoff 1.2.4  released 2019-03-12

* Let Stream#bt/#text accept a nil string


## podoff 1.2.3  released 2019-03-12

* Introduce Stream#rg(red, green, blue)
* Silence 2 Ruby warnings


## podoff 1.2.2  released 2017-02-13

* Use rb: + enc instead of just r: + enc, should fix issues on Windows


## podoff 1.2.1  released 2017-02-01

* Fail with ArgumentError when pdf not "unpacked"


## podoff 1.2.0  released 2015-11-11

* require encoding upon loading and parsing, introduce Document#encoding
* drop Podoff::Obj#page_number
* use /Kids in /Pages to determine pages and page order


## podoff 1.1.1  released 2015-10-26

* reworked xref table output
* FlateDecode stream if length > 98


## podoff 1.1.0  released 2015-10-25

* more tolerant at parsing (StringScanner)
* bin/podoff
* Document#rewrite(path)


## podoff 1.0.0  released 2015-10-23

* leverage incremental updates


## podoff 0.9.1  not released

* ensure Obj#contents accepts arrays


## podoff 0.9.0  released 2015-10-21

* beta release


## podoff 0.0.1  released 2015-10-20

* initial, empty, release



spec/obj_spec.rb

#
# specifying podoff
#
# Tue Oct 20 15:08:59 JST 2015
#

require 'spec_helper'


describe Podoff::Obj do

  before :all do

    @d = Podoff.load('pdfs/udocument0.pdf', 'iso-8859-1')
  end

  describe '#document' do

    it 'points to the Podoff::Document owning this Obj' do

      expect(@d.objs.values.first.document).to eq(@d)
    end
  end

  describe '#source' do

    it 'returns the source behind the obj' do

      o = @d.objs['20 0']

      expect(o.source).to eq(%{
20 0 obj
<< /DA (/Calibri,Bold 10 Tf 0 g) /F 4 /FT /Tx /MK << >> /P 58 0 R /Rect [ 448.723 652.574 490.603 667.749 ] /Subtype /Widget /T (State) /TU (State) /Type /Annot >>
endobj
      }.strip)
    end
  end

#  describe '#match' do
#
#    it 'returns a MatchData instance if there is a match' do
#
#      o = @d.objs['1 0']
#
#      m = o.match(/\/Contents ([^\n]+)/)
#
#      expect(m).not_to eq(nil)
#      expect(m[1]).to eq('3 0 R')
#      expect(m.offset(0)).to eq([ 123, 138 ]) # /!\
#    end
#
#    it 'returns nil if the match exits the obj' do
#
#      o = @d.objs['1 0']
#
#      m = o.match(/3 0 obj/)
#
#      expect(m).to eq(nil)
#    end
#
#    it 'returns nil if there is no match' do
#
#      o = @d.objs['1 0']
#
#      m = o.match(/nada/)
#
#      expect(m).to eq(nil)
#    end
#  end
#
#  describe '#dmatch' do
#
#    it 'matches with the zero offset set to the document' do
#
#      o = @d.objs['1 0']
#
#      m = o.dmatch(/\/Contents ([^\n]+)/)
#
#      expect(m).not_to eq(nil)
#      expect(m[1]).to eq('3 0 R')
#      expect(m.offset(0)).to eq([ 138, 153 ]) # /!\
#    end
#  end

  describe '#type' do

    it 'returns the type of the obj' do

      expect(@d.objs['12 0'].type).to eq('/Font')
    end

    it 'returns nil if there is no type' do

      expect(@d.objs['59 0'].type).to eq(nil)
    end

    it 'works on open streams' do

      st = @d.add_stream

      expect(st.obj.type).to eq(nil)
    end
  end

#  describe '#parent' do
#
#    it 'returns the parent ref if any' do
#
#      expect(@d.objs.values.first.parent).to eq('2 0')
#    end
#
#    it 'returns nil if there is no parent' do
#
#      expect(@d.objs['2 0'].parent).to eq(nil)
#    end
#  end

#  describe '#kids' do
#
#    it 'returns a list of refs' do
#
#      expect(@d.objs['2 0'].kids).to eq([ '1 0', '16 0', '33 0' ])
#    end
#
#    it 'returns an empty list if there are no kids' do
#
#      expect(@d.objs['224 0'].kids).to eq([])
#    end
#  end
#
#  describe '#contents' do
#
#    it 'returns the Contents references (single)' do
#
#      expect(@d.objs['1 0'].contents).to eq([ '3 0' ])
#    end
#
#    it 'returns the Contents references (array)' do
#
#      expect(@d.objs['16 0'].contents).to eq([ '17 0' ])
#    end
#
#    it 'returns an empty list if none' do
#
#      expect(@d.objs['224 0'].contents).to eq([])
#    end
#  end

  context 'insertions' do

    before :each do

      @d = Podoff.load('pdfs/udocument0.pdf', 'iso-8859-1')
    end

    describe '#insert_contents' do

      it 'fails if the target hasn\'t been replicated' do

        expect {
          @d.objs['23 0'].insert_contents('-1 0')
        }.to raise_error(ArgumentError, "target '23 0' not a replica")
      end

      it 'fails if the target doesn\'t have /Contents' do

        expect {
          ta = @d.re_add('23 0')
          ta.insert_contents('-1 0')
        }.to raise_error(ArgumentError, "target '23 0' doesn't have /Contents")
      end

      it 'accepts an obj' do

        pa = @d.re_add(@d.page(1))

        st = @d.add_stream('BT 70 80 Td /Font0 35 Tf (content is king!) Tj ET')

        pa.insert_contents(st)

        expect(pa.source).to match(/\/Contents \[151 0 R #{st.ref} R\]/)
      end

      it 'accepts an obj ref' do

        pa = @d.re_add(@d.page(1))

        st = @d.add_stream('BT 70 80 Td /Font0 35 Tf (content is king!) Tj ET')

        pa.insert_contents(st.ref)

        expect(pa.source).to match(/\/Contents \[151 0 R #{st.ref} R\]/)
      end
    end

    describe '#insert_font' do

      it 'fails if the target hasn\'t been replicated' do

        expect {
          @d.objs['23 0'].insert_font('/Helvetica', '-1 0')
        }.to raise_error(ArgumentError, "target '23 0' not a replica")
      end

      it 'accepts name, obj' do

        fo = @d.add_base_font('/Helvetica')
        pa = @d.re_add(@d.page(1))

        pa.insert_font('MyHelv', fo)

        expect(pa.source).to match(/\/Font\s+<<\s+\/MyHelv #{fo.ref} R\s+/)
      end

      it 'accepts name, obj ref' do

        fo = @d.add_base_font('/Helvetica')
        pa = @d.re_add(@d.page(1))

        pa.insert_font('MyHelv', fo.ref)

        expect(pa.source).to match(/\/Font\s+<<\s+\/MyHelv #{fo.ref} R\s+/)
      end

      it 'accepts a slash in front of the name' do

        fo = @d.add_base_font('/Helvetica')
        pa = @d.re_add(@d.page(1))

        pa.insert_font('/MyHelv', fo.ref)

        expect(pa.source).to match(/\/Font\s+<<\s+\/MyHelv #{fo.ref} R\s+/)
      end
    end

    describe '#add_to_attribute' do

      it 'adds to a list of references' do

        d = Podoff.load('pdfs/qdocument0.pdf', 'iso-8859-1')

        o = d.re_add('56 0')

        o.send(:add_to_attribute, :contents, '9999 0')

        expect(o.attributes).to eq(
          { type: '/Page', contents: '[151 0 R 9999 0 R]' })
      end
    end
  end
end



spec/core_spec.rb

#
# specifying podoff
#
# Tue Oct 20 13:11:38 JST 2015
#

require 'spec_helper'


describe Podoff do

  describe '.load' do

    it 'loads a PDF document' do

      d = Podoff.load('pdfs/t0.pdf', 'utf-8')

      expect(d.class).to eq(Podoff::Document)
      expect(d.objs.keys).to eq([ '1 0', '2 0', '3 0', '4 0', '5 0', '6 0' ])
      expect(d.xref).to eq(413)

      #pp d.objs.values.collect(&:to_a)

      expect(d.objs['1 0'].to_a).to eq(
        [ '1 0', 9, 54, { type: '/Catalog' } ])
      expect(d.objs['3 0'].to_a).to eq(
        [ '3 0', 111, 213, { type: '/Page', contents: '6 0 R' } ])

      expect(d.objs.values.first.document).not_to be(nil)

      expect(d.obj_counters.keys).to eq(
        [ '1 0', '2 0', '3 0', '4 0', '5 0', '6 0' ])
      expect(d.obj_counters.values).to eq(
        [ 1, 1, 1, 1, 1, 1 ])

      expect(d.root).to eq('1 0')

      expect(d.pages.size).to eq(1)
    end

    it 'loads a PDF document' do

      d = Podoff.load('pdfs/udocument0.pdf', 'iso-8859-1')

      expect(d.class).to eq(Podoff::Document)
      expect(d.xref).to eq(1612815)
      expect(d.objs.size).to eq(273)
      expect(d.objs.keys).to include('1 0')
      expect(d.objs.keys).to include('273 0')

      expect(d.root).to eq('1 0')

      expect(d.pages.size).to eq(3)
    end

    it 'loads a PDF document with incremental updates' do

      d = Podoff.load('pdfs/t1.pdf', 'utf-8')

      expect(d.class).to eq(Podoff::Document)
      expect(d.xref).to eq(704)
      expect(d.objs.keys).to eq([ '1 0', '2 0', '3 0', '4 0', '5 0', '6 0' ])

      expect(d.obj_counters.keys).to eq(
        [ '1 0', '2 0', '3 0', '4 0', '5 0', '6 0' ])
      expect(d.obj_counters.values).to eq(
        [ 1, 1, 1, 1, 1, 2 ])

      expect(d.root).to eq('1 0')
    end

    it 'loads a [re]compressed PDF documents' do

      d = Podoff.load('pdfs/qdocument0.pdf', 'iso-8859-1')

      expect(d.class).to eq(Podoff::Document)
      expect(d.xref).to eq(1612815)
      expect(d.objs.size).to eq(273)

      expect(d.root).to eq('1 0')

      #d.objs.each do |ref, o|
      #  p [ o.ref, o.attributes ]
      #end

      expect(d.pages.size).to eq(3)
      expect(d.objs['46 0'].attributes[:type]).to eq('/Annot')
    end

    it 'rejects items that are not PDF documents' do

      expect {
        Podoff.load('spec/spec_helper.rb', 'utf-8')
      }.to raise_error(ArgumentError, 'not a PDF file')
    end
  end
end



spec/alpha_spec.rb

#
# specifying podoff
#
# Tue Nov 10 21:01:51 JST 2015
#

require 'spec_helper'


describe 'fixtures:' do

  Dir['pdfs/*.pdf'].each do |path|

    describe path do

      it 'is a valid pdf document' do

        expect(path).to be_a_valid_pdf
      end
    end
  end

  describe 'pdfs/t0.pdf' do

    it 'is encoded as UTF-8' do

      expect('pdfs/t0.pdf').to be_encoded_as('utf-8')
    end
  end

  describe 'pdfs/udocument0.pdf' do

    it 'is encoded as ISO-8859-1' do

      expect('pdfs/udocument0.pdf').to be_encoded_as('latin1')
    end
  end
end



scratch/ss.rb

require 'strscan'

#s = File.read('/Users/jmettraux/tmp/sgpdfs/ig.pdf')
s =
  File.open('/Users/jmettraux/tmp/sgpdfs/ig.pdf', 'r:iso-8859-1') { |f|
    f.read
  }
p s.size

#keywords = %w[ obj endobj /Type stream endstream ]

#s.scan(/(endobj|obj)/) do |m|
#  p m.class
#end

ss = StringScanner.new(s)
version = ss.scan(/%PDF-\d+\.\d+/)
p version
objs = []
loop do
  m = ss.scan_until(/\d+ \d+ obj/); break unless m
  ref = m.match(/\d+ \d+/)[0]
  st = ss.pos - ref.length - 3
  m = ss.scan_until(/endobj\s/); break unless m
  en = ss.pos - 1
  objs << [ ref, st, en ]
end
p objs

class << ss
  def peekch
    c = getch
    self.pos = self.pos - 1
    c
  end
  def bkw(char)
    loop do
      self.pos = self.pos - 1
      break if self.pos < 0 || peekch == char
    end
  end
  def frw(char)
    loop do
      c = peekch
      break if c == nil || c == char
      self.pos = self.pos + 1
    end
  end
end

root = nil
ss.pos = 0
loop do
  c = ss.skip_until(/\/Root([\s0-9R]+)/)
  break unless c
  #ss.bkw('/'); ss.frw(' ')
  3.times { ss.bkw(' ') }; ss.pos += 1
  root = ss.scan(/\d+ \d+/)
end
p [ :root, root ]



lib/podoff.rb

require 'zlib'
require 'strscan'
require 'stringio'


module Podoff

  VERSION = '1.4.0'

  def self.load(path, encoding)

    Podoff::Document.load(path, encoding)
  end

  def self.parse(s, encoding)

    Podoff::Document.new(s, encoding)
  end

  class Document

    def self.load(path, encoding)

      Podoff::Document.new(
        File.open(path, 'rb:' + encoding) { |f| f.read },
        encoding
      )
    end

    def self.parse(s)

      Podoff::Document.new(s)
    end

    attr_reader :encoding

    attr_reader :scanner
    attr_reader :version
    attr_reader :xref
    attr_reader :objs
    attr_reader :obj_counters
    attr_reader :root
    #
    attr_reader :additions

    def initialize(s, encoding)

      fail ArgumentError.new('not a PDF file') \
        unless s.match(/\A%PDF-\d+\.\d+\s/)

      @encoding = encoding

      @scanner = ::StringScanner.new(s)
      @version = nil
      @xref = nil
      @objs = {}
      @obj_counters = {}
      @root = nil

      @additions = {}

      @version = @scanner.scan(/%PDF-\d+\.\d+/)

      loop do

        @scanner.skip_until(
          /(startxref\s+\d+|\d+\s+\d+\s+obj|\/Root\s+\d+\s+\d+\s+R)/)

        m = @scanner.matched
        break unless m

        if m[0] == 's'
          @xref = m.split(' ').last.to_i
        elsif m[0] == '/'
          @root = extract_ref(m)
        else
          obj = Podoff::Obj.extract(self)
          @objs[obj.ref] = obj
          @obj_counters[obj.ref] = (@obj_counters[obj.ref] || 0) + 1
        end
      end

      if @root == nil
        @scanner.pos = 0
        loop do
          @scanner.skip_until(/\/Root\s+\d+\s+\d+\s+R/)
          break unless @scanner.matched
          @root = extract_ref(@scanner.matched)
        end
      end
    end

    def source

      @scanner.string
    end

    def updated?

      @additions.any?
    end

    def dup

      o = self

      self.class.allocate.instance_eval do

        @encoding = o.encoding

        @scanner = ::StringScanner.new(o.source)
        @xref = o.xref

        @objs = o.objs.inject({}) { |h, (k, v)| h[k] = v.dup(self); h }
        @obj_counters = o.obj_counters.dup

        @root = o.root

        @additions =
          o.additions.inject({}) { |h, (k, v)| h[k] = v.dup(self); h }

        self
      end
    end

    def pages

      #@objs.values.select { |o| o.type == '/Page' }

      ps = @objs.values.find { |o| o.type == '/Pages' }

      fail ArgumentError.new(
        "no /Pages, the PDF is not usable by Podoff as is, you have to do " +
        "`qpdf --object-streams=disable original.pdf unpacked.pdf` " +
        "and use unpacked.pdf instead of original.pdf"
      ) unless ps

      extract_refs(ps.attributes[:kids]).collect { |r| @objs[r] }
    end

    def page(index)

      if index < 0
        pages[index]
      elsif index == 0
        nil
      else
        pages[index - 1]
      end
    end

    def new_ref

      "#{
        @objs.keys.inject(-1) { |i, r| [ i, r.split(' ').first.to_i ].max } + 1
      } 0"
    end

    def add(obj)

      @objs[obj.ref] = obj
      @additions[obj.ref] = obj

      obj
    end

    def add_base_font(name)

      name = name[1..-1] if name[0] == '/'

      r = new_ref
      s = "#{r} obj <</Type /Font /Subtype /Type1 /BaseFont /#{name}>> endobj"

      add(Obj.new(self, r, source: s))
    end

    def add_stream(src=nil, &block)

      ref = new_ref

      src =
        src &&
        [
          "#{ref} obj",
          "<< /Length #{src.size} >>\nstream\n#{src}\nendstream",
          "endobj"
        ].join("\n")

      str =
        src ?
        nil :
        make_stream(&block)

      obj = add(Obj.new(self, ref, source: src, stream: str))

      str || obj
    end

    def re_add(obj_or_ref)

      obj = obj_or_ref.is_a?(String) ? @objs[obj_or_ref] : obj_or_ref

      obj = obj.replicate unless obj.replica?

      add(obj)
    end

    def write(path=:string, encoding=nil)

      encoding ||= @encoding

      f =
        case path
          when :string, '-' then StringIO.new
          when String then File.open(path, 'wb')
          else path
        end
      f.set_encoding(encoding) # internal encoding: nil
      #f.set_encoding(encoding, encoding)

      f.write(source)

      if @additions.any?

        pointers = {}

        @additions.values.each do |o|
          f.write("\n")
          pointers[o.ref.split(' ').first.to_i] = f.pos
          f.write(o.to_s.force_encoding(encoding))
        end
        f.write("\n\n")

        xref = f.pos

        write_xref(f, pointers)

        f.write("trailer\n")
        f.write("<<\n")
        f.write("/Prev #{self.xref}\n")
        f.write("/Size #{objs.size + 1}\n")
        f.write("/Root #{root} R\n")
        f.write(">>\n")
        f.write("startxref #{xref}\n")
        f.write("%%EOF\n")
      end

      f.close if path.is_a?(String) || path.is_a?(Symbol)

      f.is_a?(StringIO) ? f.string : nil
    end

    def rewrite(path=:string, encoding=nil)

      encoding ||= @encoding

      f =
        case path
          when :string, '-' then StringIO.new
          when String then File.open(path, 'wb')
          else path
        end
      f.set_encoding(encoding)

      v = source.match(/%PDF-\d+\.\d+/)[0]
      f.write(v)
      f.write("\n")

      pointers = {}

      objs.keys.sort.each do |k|
        pointers[k.split(' ').first.to_i] = f.pos
        f.write(objs[k].source.force_encoding(encoding))
        f.write("\n")
      end

      xref = f.pos

      write_xref(f, pointers)

      f.write("trailer\n")
      f.write("<<\n")
      f.write("/Size #{objs.size + 1}\n")
      f.write("/Root #{root} R\n")
      f.write(">>\n")
      f.write("startxref #{xref}\n")
      f.write("%%EOF\n")

      f.close if path.is_a?(String) || path.is_a?(Symbol)

      f.is_a?(StringIO) ? f.string : nil
    end

    protected

    def write_xref(f, pointers)

      f.write("xref\n")
      f.write("0 1\n")
      f.write("0000000000 65535 f \n")

      pointers
        .keys
        .sort
        .inject([ [] ]) { |ps, k|
          ps << [] if ps.last != [] && k > ps.last.last + 1
          ps.last << k
          ps
        }
        .each { |part|
          f.write("#{part.first} #{part.size}\n")
          part.each { |k| f.write(sprintf("%010d 00000 n \n", pointers[k])) }
        }
    end

    def make_stream(&block)

      s = Stream.new
      s.instance_exec(&block) if block

      s
    end

    def extract_ref(s)

      s.gsub(/\s+/, ' ').gsub(/[^0-9 ]+/, '').strip
    end

    def extract_refs(s)

      s.gsub(/\s+/, ' ').scan(/(\d+ \d+) R/).collect(&:first)
    end
  end

  class Obj

    ATTRIBUTES = { type: 'Type', contents: 'Contents', kids: 'Kids' }

    def self.extract(doc)

      sca = doc.scanner

      re = sca.matched[0..-4].strip
      st = sca.pos - sca.matched.length

      i = sca.skip_until(/endobj/); return nil unless i
      en = sca.pos - 1

      Podoff::Obj.new(doc, re, start_index: st, end_index: en)
    end

    attr_reader :document
    attr_reader :ref
    attr_reader :start_index, :end_index
    attr_reader :stream
    attr_reader :attributes

    def initialize(doc, ref, opts={})

      @document = doc
      @ref = ref

      @start_index = opts[:start_index]
      @end_index = opts[:end_index]
      @attributes = nil
      @source = opts[:source]

      @stream = opts[:stream]
      @stream.obj = self if @stream

      recompute_attributes
      #@source.obj = self if @source.is_a?(Podoff::Stream)

      @document.scanner.pos = @end_index if @document.scanner && @end_index
    end

    def dup(new_doc)

      self.class.new(
        new_doc, ref,
        start_index: start_index, end_index: end_index)
    end

    #def self.create(doc, ref, source)
    #  self.new(doc, ref, nil, nil, nil, source)
    #end

    def replicate

      self.class.new(document, ref, source: source.dup)
    end

    def to_a

      [ @ref, @start_index, @end_index, @attributes ]
    end

    def source

      @source || (@start_index && @document.source[@start_index..@end_index])
    end

    def replica?

      @source != nil
    end

    def type

      @attributes && @attributes[:type]
    end

    def insert_font(nick, obj_or_ref)

      fail ArgumentError.new("target '#{ref}' not a replica") \
        unless @source

      nick = nick[1..-1] if nick[0] == '/'

      re = obj_or_ref
      re = re.ref if re.respond_to?(:ref)

      @source = @source.gsub(/\/Font\s*<</, "/Font\n<<\n/#{nick} #{re} R")
    end

    def insert_contents(obj_or_ref)

      fail ArgumentError.new("target '#{ref}' not a replica") \
        unless @source
      fail ArgumentError.new("target '#{ref}' doesn't have /Contents") \
        unless @attributes[:contents]

      re = obj_or_ref
      re = re.obj if re.respond_to?(:obj) # Stream
      re = re.ref if re.respond_to?(:ref)

      add_to_attribute(:contents, re)
    end
    alias insert_content insert_contents

    def to_s

      source || stream.to_s
    end

    protected

    def recompute_attributes

      st, en, sca =
        if @start_index
          [ @start_index, @end_index, @document.scanner ]
        elsif @source
          [ 0, @source.length, ::StringScanner.new(@source) ]
        end

      return unless sca

      @attributes =
        ATTRIBUTES.inject({}) do |h, (k, v)|
          sca.pos = st
          i = sca.skip_until(/\/#{v}\b/)
          h[k] = sca.scan(/ *\/?[^\n\r\/>]+/).strip if i && sca.pos < en
          h
        end
    end

    def concat(refs, ref)

      refs = refs.strip
      refs = refs[1..-2] if refs[0] == '['

      "[#{refs} #{ref} R]"
    end

    def add_to_attribute(key, ref)

      fail ArgumentError.new("obj not replicated") unless @source

      pkey = ATTRIBUTES[key]

      if v = @attributes[key]
        v = concat(v, ref)
        @source = @source.gsub(/#{pkey} ([\[\]0-9 R]+)/, "#{pkey} #{v}")
      else
        i = @source.index('/Type ')
        @source.insert(i, "/#{pkey} [#{ref} R]\n")
      end
      recompute_attributes
    end
  end

  class Stream

    attr_accessor :obj
    attr_accessor :deflate

    def initialize(obj=nil, opts={})

      @obj = obj
      @font = nil
      @color = nil
      @content = StringIO.new
      @deflate = opts.has_key?(:deflate) ? opts[:deflate] : true
    end

    def tf(font_name, font_size)

      n = font_name[0] == '/' ? font_name[1..-1] : font_name

      @font = "/#{n} #{font_size} Tf "
    end
    alias font tf

    def rg(*a)

      @color = to_rg(a)
    end
    alias color rg
    alias rgb rg

    def bt(x, y, text, opts={})

      return unless text

      rgb = opts[:rgb]

      @content << "\n" if @content.size > 0
      @content << 'BT '
      @content << @font if @font
      if rgb
        @content << to_rg(rgb)
      elsif @color
        @content << @color
      end
      @content << "#{x} #{y} Td (#{escape(text)}) Tj"
      @content << ' ET'
    end
    alias text bt

    def write(text)

      @content.write(text)
    end

      #def re(x, y, w, h, opts={})
      #def re([ x, y ], [ w, h ], opts={})
      #def re([ x, y ], opts)
      #
    def re(x, *a)

      a = [ x, a ].flatten
      opts = a.last.is_a?(Hash) ? a.pop : {}
      x = a.shift; y = a.shift

      rgb = opts[:rgb]
      w = opts[:width] || opts[:w] || a[0]
      h = opts[:height] || opts[:h] || a[1]

      @content << "\n" if @content.size > 0
      @content << to_rg(rgb) if rgb
      @content << lineup(x, y, w, h) << ' re f'
    end
    alias rect re
    alias rectangle re

      #def line(x0, y0, x1, y1, x2, y2, ..., opts={})
      #def line([ x0, y0 ], [ x1, y1 ], [ x2, y2 ], ..., opts={})
      #
    def line(x0, y0, *a)

      a = [ x0, y0, a ].flatten
      opts = a.last.is_a?(Hash) ? a.pop : {}
      x0, y0, *xys = a

      rgb = opts[:rgb] || opts[:rg]
      w = opts[:width] || opts[:w]

      @content << "\n" if @content.size > 0
      @content << w.to_s << ' w ' if w
      @content << to_rg(rgb) if rgb
      @content << lineup(x0, y0) << ' m '
      xys.each_slice(2) { |x, y|
        @content << lineup(x, y, 'l ') }
      @content << 'h S'
    end

    def to_s

      s = @content.string
      f = ''
      if @deflate && s.length > 98
        f = ' /Filter /FlateDecode'
        s = Zlib::Deflate.deflate(s)
      end

      "#{obj.ref} obj\n" +
      "<</Length #{s.size}#{f}>>\nstream\n#{s}\nendstream\n" +
      "endobj"
    end

    protected

    def escape(s); s.gsub(/\(/, '\(').gsub(/\)/, '\)'); end
    def lineup(*a); a.flatten.collect(&:to_s).join(' '); end

    COLORS = {
      'black' => [ 0.0, 0.0, 0.0 ],
      'white' => [ 1.0, 1.0, 1.0 ],
      'red' => [ 1.0, 0.0, 0.0 ],
      'green' => [ 0.0, 1.0, 0.0 ],
      'blue' => [ 0.0, 0.0, 1.0 ] }

    def to_rg(a)

      a = a[0].to_s if a.length == 1

      lineup(
        if a.is_a?(Array) && a.length == 3
          a
        elsif a.is_a?(String) && a.match(/^#?([0-9a-z]{3}|[0-9a-z]{6})$/i)
          hex_to_rgb(a)
        else
          COLORS[a] || COLORS['red'] # else, stand out in RED
        end,
        'rg ')
    end

    def hex_to_rgb(s)

      s = s[1..-1] if s[0, 1] == '#'

      s.chars
        .each_slice(
          s.length == 6 ? 2 : 1)
        .collect { |x|
          sprintf('%0.4f', (x.join.to_i(16) / 255.0)).gsub(/0+$/, '0') }
    end
  end
end

